wgpu-core/src/weak_vec.rs

```diff
@@ -6,57 +6,44 @@ use std::sync::Weak;
 /// dropping older elements that no longer have strong references to them.
 #[derive(Debug)]
 pub(crate) struct WeakVec<T> {
-    inner: Vec<Option<Weak<T>>>,
-    empty_slots: Vec<usize>,
-    scan_slots_on_next_push: bool,
+    inner: Vec<Weak<T>>,
 }
 
 impl<T> Default for WeakVec<T> {
     fn default() -> Self {
         Self {
             inner: Default::default(),
-            empty_slots: Default::default(),
-            scan_slots_on_next_push: false,
         }
     }
 }
 
 impl<T> WeakVec<T> {
     pub(crate) fn new() -> Self {
-        Self {
-            inner: Vec::new(),
-            empty_slots: Vec::default(),
-            scan_slots_on_next_push: false,
-        }
+        Self { inner: Vec::new() }
     }
 
     /// Pushes a new element to this collection.
     ///
     /// If the inner Vec needs to be reallocated, we will first drop older elements that
     /// no longer have strong references to them.
     pub(crate) fn push(&mut self, value: Weak<T>) {
-        if self.scan_slots_on_next_push {
-            for (i, value) in self.inner.iter_mut().enumerate() {
-                if let Some(w) = value {
-                    if w.strong_count() == 0 {
-                        *value = None;
-                        self.empty_slots.push(i);
-                    }
+        if self.inner.len() == self.inner.capacity() {
+            let mut i = 0;
+            while let Some(w) = self.inner.get(i) {
+                if w.strong_count() == 0 {
+                    self.inner.swap_remove(i);
+                } else {
+                    i += 1;
                 }
```

Comment:
suggestion(non-blocking): (CC @jimblandy):

```suggestion
            self.inner.retain(|w| w.strong_count() > 0);
```

It's not clear-cut that `swap_remove` is going to be faster, and this is a lot simpler code to read. See also discussion from <https://github.com/rust-lang/rust/issues/69043>.

I guess I'd suggest profiling, making sure folks are happy (or at least happier), and then trying `retain` as an iteration on top.

Reply:
With `retain` I'm concerned about the nr of moves, if one of the elements close to the front gets dropped, all of the elements after it have to be moved. `retain` will take better advantage of caches though.

Ideally optimized apps don't create and destroy resources often but depending on how they create and destroy resources one solution might be faster than the other.

A perf cliff of `swap_remove` would be if all resources have been dropped...

Profiling multiple scenarios would be best.

Reply:
We don't have to block merging on profiling here, OFC, just wanted to note that profiling and _then_ applying the change seems even like a `good-first-issue` task.

Reply:
I think it's good to bound the number of moves on the number of broken weak references, since we know work will be proportional to at least that in any implementation.